numpy/core/src/multiarray/methods.c

```diff
@@ -619,10 +619,9 @@ array_tofile(PyArrayObject *self, PyObject *args, PyObject *kwds)
     if (fd == NULL) {
         goto fail;
     }
-    if (PyArray_ToFile(self, fd, sep, format) < 0) {
-        goto fail;
-    }
-    if (npy_PyFile_DupClose2(file, fd, orig_pos) < 0) {
+    res1 = PyArray_ToFile(self, fd, sep, format);
+    res2 = npy_PyFile_DupClose2(file, fd, orig_pos);
+    if (res1 < 0 || res2 < 0) {
         goto fail;
     }
```

Comment:
I suspect this might be unsafe - if the first throws an exception, it's not safe to call back into python code before catching it - and `npy_PyFile_DupClose2` looks like it can call back in. You might need to use
```suggestion
    int res_write = PyArray_ToFile(self, fd, sep, format);
    {
        PyObject *type, *value, *traceback;
        PyErr_Fetch(&type, &value, &traceback);
        int res_close = npy_PyFile_DupClose2(file, fd, orig_pos);
        npy_PyErr_ChainExceptions(type, value, traceback);
        if (res_write < 0 || res_close < 0) {
            return NULL;
        }
    }
```
which also means you'll get chained exceptions